### PR TITLE
Add `Phi-3-mini-4k-instruct` checkpoint

### PR DESCRIPTION
**Pull request by rasbt**
- [x] Verify Phi-3-mini-4k-instruct configs
- [x] Add prompt style
- [x] ~Add other config files~ (will be added in a separate PR)
- [x] Add test_model.py
- [x] Add support in convert_hf_checkpoint.py
- [x] Add support in convert_lit_checkpoint.py
- [x] Add tests for conversion scripts
- [x] Add to test_prompts.py
- [x] Update 2 tables in README
- [x] Update download_model_weights.md

### CODE REVIEW

**Comment by Andrei-Aksionov** [comment]
There is a modeling_*.py file.
Good luck 🙂.

**Comment by rasbt** [comment]
> There is a modeling_*.py file.
> Good luck 🙂.

Haha, I finally get the weights loaded but of course it's never easy ... of course it's generating gibberish 

```
⚡ phi-3-checkpoint ~/litgpt litgpt chat --checkpoint_dir checkpoints/microsoft/Phi-3-mini-4k-instruct
Now chatting with Phi-3-mini-4k-instruct.
To exit, press 'Enter' on an empty prompt.

Seed set to 1234
>> Prompt: What do llamas eat?
>> Reply: epsonniformes }).selves }).SSIONunicívo }). EverythingFormsћassaiejalphutureievediennesenticaciónicaciónMilMinigh ninassaselvesselves exhaustselvesonnselvesktionΗracheracheionedΗ Avenoted Bij_+versionsmastevosepsselvesmobileselvesilleryassaucealphasseestoreselvesférFormsiej Mu Kaiser oppienngnatteversionsionedionedversionsSSIONectionaccoossFormassaselves_+uminatesonoSSIONológissancecenteecause_+ienn选uraleʋ Stepalphigosionaliilonverte }).ienn }).ativo Sternsonoiejuralassawnkademselves│uraleativaionedvos_+utschversionsponiej_+icacióniejiewerológvoasonverte shoutioned位ionedIdentmobi
```


Let the easter egg hunt begin 😭

**Comment by rasbt** [comment]
Some more tidbits via [Daniel Han](https://twitter.com/danielhanchen/status/1782853167572832650):

> Phi 3 (3.8B) got released! The paper said it was just a Llama arch, but I found some quirks while adding this to 
[@UnslothAI](https://twitter.com/UnslothAI)
:

1. Sliding window of 2047? Mistral v1 4096. So does Phi mini have SWA? (And odd num?) Max RoPE position is 4096?
2. Upcasted RoPE? Like Gemma?
3. Dynamic RoPE for 128K context lengths
4. Fused MLP & QKV - need to unfuse
5. MMLU evals are very different betw the Phi team Llama-3 team - why?

**Comment by Andrei-Aksionov** [comment]
Ok, it's becoming more interesting.
Somewhat I expected from LlaMA 3, but it didn't deliver.

File: litgpt/model.py [hunk 1]
```diff
@@ -298,6 +298,20 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return self.proj(x)
 
 
+class Phi3MLP(nn.Module):
```
Comment (carmocca) [inline litgpt/model.py:301]: It should be possible to not need this class at all and instead reshape the weights for LLaMAMLP in the checkpoint conversion.

Comment (rasbt) [reply litgpt/model.py:301]: 100% agree. I was thinking the same thing. Similar to OLMo, I was hoping to get it working first and then simplify from there.

Comment (Andrei-Aksionov) [reply litgpt/model.py:301]: New models by Apple have if-else statements for this case: https://huggingface.co/apple/OpenELM-270M-Instruct/blob/main/modeling_openelm.py#L405-L462

For simplicity we definitely shouldn't do the same.

**Comment by rasbt** [comment]
Looks like the sliding window number was a typo: https://huggingface.co/microsoft/Phi-3-mini-4k-instruct/commit/b043e05a86cfc77f8d53eb0edf6a33e39afbcb5e
Screenshot 2024-04-24 at 12 21 02 PM

**Comment by Andrei-Aksionov** [comment]
Current code is an ugly state, but at least the model produces the same output as HF one.
The most notable change is that now Phi3 model doesn't use `parallel_residual` in contrast to Phi1.5 and Phi2.

The missing piece is the Tokenizer: it has a smaller vocab size (32k vs 50k) that was extended by 64 special tokens. 
If I'm not mistaken, the current code doesn't add these tokens.

File: litgpt/scripts/convert_hf_checkpoint.py [hunk 2]
```diff
@@ -266,6 +279,21 @@ def copy_weights_phi(
             del qkv_weights[i][weight_type]
 
 
+def qkv_reassemble(param: Union[torch.Tensor, NotYetLoadedTensor], config: Config) -> torch.Tensor:
```
Comment (rasbt) [inline litgpt/scripts/convert_hf_checkpoint.py:282]: Thanks for the cleanup here

File: litgpt/model.py [hunk 3]
```diff
@@ -301,8 +301,8 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 class Phi3MLP(nn.Module):
     def __init__(self, config: Config) -> None:
         super().__init__()
-        self.gate_up_proj = nn.Linear(config.n_embd, config.intermediate_size, bias=config.bias)
-        self.down_proj = nn.Linear(config.intermediate_size//2, config.n_embd, bias=config.bias)
+        self.gate_up_proj = nn.Linear(config.n_embd, 2 * config.intermediate_size, bias=config.bias)
```
Comment (rasbt) [inline litgpt/model.py:304]: Huh, that's what I had originally, because it matches their description, but I got dimension mismatches with that.

**Comment by rasbt** [comment]
> The missing piece is the Tokenizer: it has a smaller vocab size (32k vs 50k) that was extended by 64 special tokens.
If I'm not mistaken, the current code doesn't add these tokens.

Yeah, that sounds about right based on the Phi-3 paper: 

> To best benefit the open source community, phi-3-mini is built upon a similar block structure as Llama-2 [TLI+23] and uses the same tokenizer with vocabulary size of 320641

File: litgpt/prompts.py [hunk 4]
```diff
@@ -272,6 +272,17 @@ def apply(self, prompt: str, **kwargs: str) -> str:
         return f"Instruct: {prompt}\nOutput:"
 
 
+class Phi3(PromptStyle):
+    def apply(self, prompt: str, **kwargs: str) -> str:
+        return (
+            "<|system|>\n"
```
Comment (carmocca) [inline litgpt/prompts.py:278]: They just changed this: https://huggingface.co/microsoft/Phi-3-mini-4k-instruct/commit/688dd7b82eb7ff6cb786a9f723ab3e35479a19d5

Comment (rasbt) [inline litgpt/prompts.py:278]:
```suggestion
            "<|user|>\n"
```

**Comment by rasbt** [comment]
A related interesting post @Andrei-Aksionov https://x.com/danielhanchen/status/1795453604532207989

**Comment by Andrei-Aksionov** [comment]
Alt Text

Required some number of changes but it works.
Also tried a quick LoRA finetune — no issues there.

@rasbt Could you check the changes in READMEs? Not 100% sure that I've done them correctly.

**Comment by rasbt** [comment]
Thanks so much! I am currently moving and offline until weekend/monday. Will take a look when I am back!

File: litgpt/scripts/convert_hf_checkpoint.py [hunk 5]
```diff
@@ -272,6 +290,24 @@ def copy_weights_phi(
...
```
Comment (rasbt) [inline litgpt/scripts/convert_hf_checkpoint.py:293]: This is great, I think this is maybe just what I needed for OLMo

File: litgpt/scripts/convert_hf_checkpoint.py [hunk 6]
```diff
@@ -321,14 +357,14 @@ def convert_hf_checkpoint(
 
     if "falcon" in model_name:
         copy_fn = partial(copy_weights_falcon, model_name)
-    elif config.mlp_class_name in ("LLaMAMLP", "GemmaMLP", "LLaMAMoE"):
+    elif model_name.startswith(("phi", "Phi")):
```
Comment (rasbt) [inline litgpt/scripts/convert_hf_checkpoint.py:360]: You could also do a `model_name.lower().startswith("phi")`, but not sure that one is better than the other. Looks fine to me as is.

File: tutorials/download_model_weights.md [hunk 7]
```diff
@@ -24,6 +24,7 @@ LitGPT supports a variety of LLM architectures with publicly available weights.
 | Nous-Hermes | 7B, 13B, 70B | NousResearch | [Org page](https://huggingface.co/NousResearch)                                                                          |
 | OpenLLaMA | 3B, 7B, 13B | OpenLM Research | [Geng & Liu 2023](https://github.com/openlm-research/open_llama)                                                         |
 | Phi | 1.3B, 2.7B | Microsoft Research  | [Li et al. 2023](https://arxiv.org/abs/2309.05463)                                                                          |
```
Comment (rasbt) [inline tutorials/download_model_weights.md:26]:
```suggestion
| Phi 1.5 & 2 | 1.3B, 2.7B | Microsoft Research  | [Li et al. 2023](https://arxiv.org/abs/2309.05463)                                                                          |
```

File: tutorials/download_model_weights.md [hunk 8]
```diff
@@ -24,6 +24,7 @@ LitGPT supports a variety of LLM architectures with publicly available weights.
...
+| Phi-3-mini | 3.8B | Microsoft Research | [Technical Report](https://arxiv.org/abs/2404.14219)
```
Comment (rasbt) [inline tutorials/download_model_weights.md:27]:
```suggestion
| Phi 3 | 3.8B | Microsoft Research | [Abdin et al. 2024](https://arxiv.org/abs/2404.14219)
```

File: litgpt/scripts/convert_lit_checkpoint.py [hunk 9]
```diff
@@ -256,11 +287,11 @@ def convert_lit_checkpoint(checkpoint_dir: Path, output_dir: Path) -> None:
 
     if "falcon" in config.name:
         copy_fn = partial(copy_weights_falcon, config.name)
+    elif config.name.startswith(("phi", "Phi")):
```
Comment (rasbt) [inline litgpt/scripts/convert_lit_checkpoint.py:290]: Same as above with `.lower()` but no change necessary.

**Review by rasbt — COMMENTED** [review]
This is a tremendous PR. Thanks so much @Andrei-Aksionov . Just tried the model and it works great!

File: litgpt/scripts/convert_hf_checkpoint.py (hunk 6 above)
Comment (rasbt) [reply litgpt/scripts/convert_hf_checkpoint.py]: Thanks for the update!

**Comment by rasbt** [comment]
I think the failing tests are because of the new Eval Harness release: https://pypi.org/project/lm-eval/#history

I can look into it in a separate PR

**Comment by Andrei-Aksionov** [comment]
Yep, this is the reason.
I "love" when bug-fix releases break code.

**Comment by rasbt** [comment]
All good now. Big thanks again @Andrei-Aksionov !!

File: litgpt/scripts/convert_hf_checkpoint.py (hunk 5 above)
Comment (aflah02) [reply litgpt/scripts/convert_hf_checkpoint.py:293]: Hi @rasbt 
I was looking for OLMo and stumbled on your comment. Are there any plans to support the OLMo architecture in the near future? Couldn't find it here - https://github.com/Lightning-AI/litgpt/blob/main/litgpt/config.py

**Pull request merged**